test/cpu_reduce_util.hpp

```
@@ -29,19 +29,196 @@
+{
+    return abs(a);
+};
+
+template <>
+inline float Abs<float>(float a)
+{
+    return fabsf(a);
+};
+
+template <>
+inline double Abs<double>(double a)
+{
+    return fabs(a);
+};
+
+template <>
+inline half_float::half Abs<half_float::half>(half_float::half a)
+{
+    return half_float::abs(a);
+};
+
+template <typename compType>
+static inline std::function<void(compType&)> PreUnaryOpFn(miopenReduceTensorOp_t op_, int divider)
+{
+    switch(op_)
+    {
+    case MIOPEN_REDUCE_TENSOR_NORM1:
+        return ([&, divider](compType& a_) {
+            a_ = Abs<compType>(a_) / convert_type<compType>(static_cast<float>(divider));
+        });
+    case MIOPEN_REDUCE_TENSOR_NORM2:
+        return ([&, divider](compType& a_) {
+            a_ = a_ * a_ / convert_type<compType>(static_cast<float>(divider));
+        });
+    case MIOPEN_REDUCE_TENSOR_AMAX: return ([&](compType& a_) { a_ = Abs<compType>(a_); });
+
+    case MIOPEN_REDUCE_TENSOR_AVG:
+    case MIOPEN_REDUCE_TENSOR_ADD:
+    case MIOPEN_REDUCE_TENSOR_MUL:
+    case MIOPEN_REDUCE_TENSOR_MIN:
+    case MIOPEN_REDUCE_TENSOR_MAX: return ([&](compType& a_) { (void)a_; });
+    }
+
+    return (std::function<void(compType&)>{});
+};
+
+template <typename compType>
+static inline std::function<void(compType&)> PosUnaryOpFn(miopenReduceTensorOp_t op_, int divider)
+{
+    (void)divider;
+    switch(op_)
+    {
+    case MIOPEN_REDUCE_TENSOR_NORM2: return ([&](compType& a_) { a_ = Sqrt<compType>(a_); });
+
+    case MIOPEN_REDUCE_TENSOR_AVG:
+        return ([&, divider](compType& a_) {
+            a_ = a_ / convert_type<compType>(static_cast<float>(divider));
+        });
+
+    case MIOPEN_REDUCE_TENSOR_ADD:
+    case MIOPEN_REDUCE_TENSOR_NORM1:
+    case MIOPEN_REDUCE_TENSOR_MUL:
+    case MIOPEN_REDUCE_TENSOR_MIN:
+    case MIOPEN_REDUCE_TENSOR_MAX:
+    case MIOPEN_REDUCE_TENSOR_AMAX: return ([&](compType& a_) { (void)a_; });
+    }
+
+    return (std::function<void(compType&)>{});
```

We should probably throw an error here.

---

Yes, Changed